scripts/pylib/twister/twisterlib/scripting.py

```diff
@@ -0,0 +1,148 @@
+    re_scenarios: list[re.Pattern] = field(init=False, default_factory=list)
+    re_platforms: list[re.Pattern] = field(init=False, default_factory=list)
+
+    # Compiles regex patterns for scenarios and platforms, and validates the element.
+    def __post_init__(self):
+        self.re_scenarios = [re.compile(pat) for pat in self.scenarios]
+        self.re_platforms = [re.compile(pat) for pat in self.platforms]
+        if not any([self.pre_script, self.post_flash_script, self.post_script]):
+            logger.error("At least one of the scripts must be specified")
+            sys.exit(1)
+        self.pre_script = self._convert_to_script(self.pre_script)
+        self.post_flash_script = self._convert_to_script(self.post_flash_script)
+        self.post_script = self._convert_to_script(self.post_script)
+
+    # Converts a dictionary to a Script instance if necessary.
+    def _convert_to_script(self, script: dict | Script | None) -> Script | None:
+        if isinstance(script, dict):
+            return Script(**script)
+        return script
+
+
+@dataclass
+# Holds a collection of scripting elements.
+class ScriptingData:
+    elements: list[ScriptingElement] = field(default_factory=list)
+
+    # Ensures all elements are ScriptingElement instances.
+    def __post_init__(self):
+        self.elements = [
+            elem if isinstance(elem, ScriptingElement) else ScriptingElement(**elem)
+            for elem in self.elements
+        ]
+
+    @classmethod
+    # Loads scripting data from a YAML file.
+    def load_from_yaml(cls, filename: Path | str, schema: dict) -> ScriptingData:
+        try:
+            raw_data = scl.yaml_load_verify(filename, schema) or []
+            return cls(raw_data)
+        except scl.EmptyYamlFileException:
+            logger.error(f'Scripting file {filename} is empty')
+            sys.exit(1)
+        except FileNotFoundError:
+            logger.error(f'Scripting file {filename} not found')
+            sys.exit(1)
+        except Exception as e:
+            logger.error(f'Error loading {filename}: {e}')
+            sys.exit(1)
+
+    # Extends the current scripting data with another set of scripting data.
+    def extend(self, other: ScriptingData) -> None:
+        self.elements.extend(other.elements)
+
+    # Finds a scripting element that matches the given scenario and platform.
+    def find_matching_scripting(self, scenario: str, platform: str) -> ScriptingElement | None:
+        matched_elements = []
+
+        for element in self.elements:
+            if not isinstance(element, ScriptingElement):
+                element = ScriptingElement(**element)
+            if element.scenarios and not _matches_element(scenario, element.re_scenarios):
+                continue
+            if element.platforms and not _matches_element(platform, element.re_platforms):
+                continue
+            matched_elements.append(element)
+
+        # Check for override_script
```

Comment:
It seems that this function allow duplicate entries all without `override` attribute,  doesn't it ? so it is still a problem with duplicates and their precedence. `Override` should help when (1) there are duplicate matches, and (2) there is only one with `override` among them to chose it. All other situations for several matched_elements (several overrides, have no overrides) are conflicts and configuration error.

Also, It is an overcomplication to have `override_script` for each of 3 scripts in an entry when duplicates are possible with (platform, scenario) key. Much easier and clear is to have the `override` attribute for an element.

Reply:
After the request to include the override_script parameter code was updated. Now, if override_script is not set to true, the script will not run. Additionally, if there are duplicates (e.g., two different scripts for the same board with override_script set to true), no script will be executed.

Reply:
To move forward I'd suggest: 1) don't use regexp matching at `_matches_element()`,  2) just fail on any duplicate found, so forget the idea of `override` attribute.
It will simplify the proposed change significantly, whereas keep most of it functionality.